### ISSUE TITLE
Henrik Sperre Johansen proposed improvements

### ISSUE THREAD

**Issue by GoogleCodeExporter** [theseion/Fuel#49]
```
Henry said:

1)
You shouldn't need more than one Cluster class for variableWord classes either, 
FLLargeNegativeIntegerCluster, FLLargePositiveIntegerCLuster, FLFloatCluster 
etc. should could all be removed in favor of a FLVariableWordClusterSerializer, 
which has the actual class it is serializing/materializing as in instvar.

serialize:
store variablesize
store named vars
store variable words

materialize:
read variablesize.
create new instance of that size.
restore named vars.
read variable words into instance in a single operation.

There's lots of mucking about to get byte ordering consistent when storing to 
file, and then restoring it when loading, look at nextWordsInto: , and 
restoreEndianness.

If stored/loaded in the platform order, you wouldn't need the restoreEndianness 
step if platform didn't change.
The way it is now, (ie. always saved in big-endian) you almost always have to 
do the step, since intel cpu's are little endian.

The following writes out a variableWord object inplatform-specific byteorder:
wideStr := 'Hello widestring (€) primitive!'.
FileStream newFileNamed: 'test.bin' do: [:fileStream |
    fileStream binary;
        nextPutAll: wideStr].

FileStream fileNamed: 'test.bin' do: [:fileStream |
    fileStream binary.
    fileStream contents]

Not sure how well it works with the buffered stream though, since nextPutAll: 
has different behaviour for streams using internal collections, you'd have to 
test if it's worth it. Either way, there shouldn't be a need for more than one 
class with serialize/materialize pseudocode as described above.

Also: For variableBytes classes, if you rewrite:

WriteStream
nextPutAll: aCollection

     | newEnd |
     collection class instSpec == aCollection class instSpec ifFalse:
         [^ super nextPutAll: aCollection ].

     newEnd := position + aCollection size.
     newEnd > writeLimit ifTrue:
         [self growTo: newEnd + 10].

     collection replaceFrom: position+1 to: newEnd  with: aCollection startingAt: 1.
     position := newEnd.

    ^ aCollection

You can now pass all variableByte classes (Like ByteString) directly to a 
stream with a ByteArray collection, and the replaceFrom:to:with:startingAt: 
primitive will work correctly, just like the file primitive does
This means you don't need special Serializers for these either,  using f.ex. 
clunky nextStringPutAll: methods with manual asByteArray conversions.

byteStr := 'Hello primitive!'.

FileStream newFileNamed: 'test.bin' do: [:fileStream |
    fileStream binary;
        nextPutAll: byteStr].

internalStream := #[] writeStream.
internalStream nextPutAll: byteStr.

FileStream fileNamed: 'test.bin' do: [:fileStream |
    fileStream binary.
    fileStream contents = internalStream contents]


TLDR; Try not to use iteration reading in single elements of a variable object, 
but rather use the batch processing methods provided by streams, which are 
already (well, in some cases at least) optimized. Fix bugs in those existing 
ones instead :)

Cheers,
Henry


------

for mor detais read thread with subject: "FUEL improvements [WAS] Re: 
Increasing the performances of a Seaside application"
```

Original issue reported on code.google.com by `marianopeck` on 19 Jun 2011 at 1:11
